storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java

```diff
@@ -0,0 +1,673 @@
+        assertThrows(RemoteResourceNotFoundException.class, () -> tieredStorage.fetchIndex(metadata, PRODUCER_SNAPSHOT));
+
+        try {
+            assertArrayEquals(new byte[0], remoteStorageVerifier.readFully(tieredStorage.fetchIndex(metadata, TRANSACTION)));
+        } catch (Exception ex) {
+            fail("Shouldn't have thrown an exception when optional file doesn't exists in the remote store");
+        }
+    }
+
+    @Test
+    public void assertStartAndEndPositionConsistency() {
+        final RemoteLogSegmentMetadata metadata = newRemoteLogSegmentMetadata(newRemoteLogSegmentId());
+
+        assertThrows(IllegalArgumentException.class, () -> tieredStorage.fetchLogSegment(metadata, -1, Integer.MAX_VALUE));
+        assertThrows(IllegalArgumentException.class, () -> tieredStorage.fetchLogSegment(metadata, 1, -1));
+        assertThrows(IllegalArgumentException.class, () -> tieredStorage.fetchLogSegment(metadata, 2, 1));
+    }
+
+    private RemoteLogSegmentMetadata newRemoteLogSegmentMetadata(final RemoteLogSegmentId id) {
+        return new RemoteLogSegmentMetadata(id, 0, 0, -1L, -1, 1000L,
+                1024, Collections.singletonMap(0, 0L));
+    }
+
+    private RemoteLogSegmentId newRemoteLogSegmentId() {
+        return new RemoteLogSegmentId(topicIdPartition, Uuid.randomUuid());
+    }
+
+    private static List<ByteBuffer> extractRecordsValue(
+            final LocalTieredStorageSnapshot snapshot,
+            final RemoteLogSegmentId id) throws IOException {
+
+        final FileRecords records = FileRecords.open(snapshot.getFile(id, SEGMENT));
+        final List<ByteBuffer> buffers = new ArrayList<>();
+
+        for (Record record: records.records()) {
+            buffers.add(record.value());
+        }
+
+        return buffers;
+    }
+
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd-HH:mm:ss");
+
+    private String generateStorageId(String testName) {
+        return format("%s-%s-%s",
+                getClass().getSimpleName(), testName, DATE_TIME_FORMATTER.format(LocalDateTime.now()));
+    }
+
+    public final class Verifier {
+        private final LocalTieredStorage remoteStorage;
+        private final TopicIdPartition topicIdPartition;
+
+        public Verifier(final LocalTieredStorage remoteStorage, final TopicIdPartition topicIdPartition) {
+            this.remoteStorage = requireNonNull(remoteStorage);
+            this.topicIdPartition = requireNonNull(topicIdPartition);
+        }
+
+        private List<Path> expectedPaths(final RemoteLogSegmentId id) {
+            final String rootPath = getStorageRootDirectory();
+            TopicPartition tp = topicIdPartition.topicPartition();
+            final String topicPartitionSubpath = format("%s-%d-%s", topicIdPartition.topicId(),
+                   tp.partition(), tp.topic());
```

Comment:
In case we want to adopt the new structure: 

```suggestion
            final String topicPartitionSubpath = format("%s-%d-%s",
                    tp.topic(), tp.partition(), topicIdPartition.topicId());
```